### ISSUE TITLE
Investigate MioTTS + MioCodec as a Japanese engine

### ISSUE THREAD

**Issue by m96-chan** [m96-chan/voxshot#106]
Companion to #105. Both are Japanese-capable zero-shot cloning models; they fail and succeed in different places, and MioTTS looks like the better bet on every axis except one.

**A spike, like #105.** But the risky part here is different, and the order of work matters — see below.

## Why this looks better than Sarashina

| | Chatterbox (shipped) | Sarashina (#105) | MioTTS 0.6B |
| --- | --- | --- | --- |
| License | MIT-ish weights | **NonCommercial** | **Apache-2.0** (codec MIT) |
| Download | ~1.5 GB | 4.1 GB | ~1.69 GB |
| ONNX sessions | 4 | 6+ | **2** |
| Runtime support | transformers.js | **needs onnxruntime-genai** | **`qwen3` already in transformers.js** |
| ONNX export exists | yes | yes (third-party) | **no** |

Two facts carry most of the weight.

**The licenses are usable.** MioTTS-0.6B is Apache-2.0 (Qwen3-0.6B-Base) and MioCodec is MIT. #105 stops dead on a NonCommercial agreement; this does not. Note the family is mixed — 0.1B is under the Falcon license and 0.4B / 1.2B / 2.6B under LFM Open License v1.0 — so **0.6B and 1.7B are the two Apache-2.0 rungs** and are the ones worth looking at.

**`qwen3` is already in `transformers.js`** (`"qwen3"` appears in the shipped 4.2.0 bundle's model-type map). #105's blocking unknown is whether its LLM can run without `onnxruntime-genai`; here the equivalent question is already answered for the LLM half.

## What it is

Qwen3-0.6B fine-tuned for TTS, ~100,000 hours of English and Japanese, emitting **content tokens at 25 Hz** into MioCodec, which decodes straight to waveform — the codec has an integrated iSTFTHead, so there is no separate vocoder. Zero-shot voice cloning from a short reference.

```
reference.wav ─▶ MioCodec encoder ─▶ global embedding ("how") + content tokens
text          ─▶ Qwen3-0.6B       ─▶ content tokens ("what", 25 Hz)
                 MioCodec decoder ─▶ waveform (24 kHz)
```

That is `embed()` and `synthesize()` with nothing left over, and 24 kHz already matches what `ChatterboxEngine` produces.

**25 Hz is the number to notice.** Chatterbox costs ~2.4 tokens per character — a 120-character chunk asks for 331 tokens (`chatterbox-engine.ts:53-77`). MioCodec is 25 tokens per *second of audio*, flat. That changes generation cost, and it changes the upstream KV-cache leak (#76), which scales with the token budget.

Upstream reports RTF 0.055–0.065 for 0.6B. **That is vLLM on an RTX 5090 and cannot be compared to our numbers** — ours are onnxruntime-web. It is a reason to measure, not a result.

## The blocker, and why it is the opposite shape to #105

**Neither model has an ONNX export.** A search for `MioTTS onnx` and `MioCodec onnx` returns nothing.

For the LLM that is routine — Qwen3 is a standard architecture with an existing conversion path, and transformers.js already knows the type.

**The codec is where this can die.** MioCodec ships only `config.yaml` and `model.safetensors` — no `config.json`, so it is not an `AutoModel`; it needs the [Kanade-Tokenizer](https://github.com/frothywater/kanade-tokenizer) implementation. And its decoder is an **iSTFTHead**. ONNX exporters cannot handle complex tensors: the Sarashina export had to replace `torch.stft` / `torch.istft` with real-valued conv/matmul equivalents to get through, and the same work is likely needed here.

So the sequencing is the point: **export the codec first.** If it cannot be exported, nothing else matters, and the LLM conversion — the easy, satisfying part — would be wasted effort. #105 has the same discipline for a different reason.

## Spike scope, in order

1. **Export MioCodec's decoder to ONNX.** Encoder too, if `embed()` is to run in the browser rather than being precomputed. Expect to deal with iSTFT.
2. If that works: export MioTTS-0.6B (Apache-2.0 rung) and confirm `transformers.js` loads it as `qwen3`.
3. Measure one Japanese utterance end to end in Node — load time, synthesis time, RTF — and put it beside the Chatterbox figures in the README table.
4. Write up what a `MioEngine` would cost: session count, what tokenization needs porting, whether `embed()` can run client-side.

## Notes

`~/.cache/huggingface` already holds `MioTTS-2.6B`, `MioTTS-GGUF` and `MioCodec-25Hz-44.1kHz-v2` — but 2.6B is LFM-licensed and 44.1 kHz does not match the 24 kHz pipeline. The spike should pull `MioTTS-0.6B` and `MioCodec-25Hz-24kHz` rather than reuse what is there.

## Out of scope

Implementing the engine. Choosing between this and Chatterbox. Anything about #105 — the two are independent, and this one is likely to be decided first.

## Done when

- [ ] It is known whether MioCodec's decoder can be exported to ONNX, with the iSTFT question answered concretely
- [ ] If it can, one measured Japanese utterance exists with numbers rather than estimates
- [ ] A follow-up issue exists with real scope, or this closes with the reason it will not proceed

**Comment by m96-chan** [comment]
## 調査結論 (2026-08-21)

**MioTTS + MioCodec は使える。** ブラウザ内で text → speech + ゼロショット声クローンまで動作し、RTX 5090 で実測した。#116 / #118 / #121 として main にマージ済み。

### 動くもの

`examples/mio-tts.html` — テキストを入れると音声が鳴り、参照音声を選べばその声になる。全部ブラウザ内 (WebGPU)、サーバ推論なし。

| | 実測 (RTX 5090) |
| --- | ---: |
| LM 生成 (q8, WebGPU 常駐) | 209 tok/s |
| テキスト → 3.48 秒の音声 | 合計 1.80 秒 (**RTF 0.52**) |
| デコーダ単体 | RTF 0.32 |
| 参照音声 → 声の embedding | 1.2 秒 / クリップ (1 回のみ) |
| コールドスタート | +4 秒 (q8 583MB + codec 523MB) |

起票時の想定どおり **ONNX なしで通った**。web-xpu-ops の op と WGSL カーネルで全段を表現でき、上流に欠けていた op はデコーダの `padding="same"` iSTFT だけ (web-xpu-ops#92 で解決済み)。エンコーダ側は欠落ゼロだった。

### 起票時の見積りとの差分

- **ONNX sessions 2 → 0**。transformers.js も onnxruntime も使っていない。q8 に変換した safetensors を直接読んで自前カーネルで回す形になった
- **ダウンロード ~1.69GB → 1.1GB** (q8 583MB + codec 523MB)。bf16 1.2GB を per-row absmax int8 に落とした。web-xpu-ops が持つ唯一の量子化形式で、fused GEMV (matvecQ8) がある
- **0.6B を採用**。Apache-2.0 の下側の段。2.6B は LFM2 アーキテクチャで別物、GGUF の 0.1B/1.2B は license か速度で見送り

### 品質の担保

各段が一段下のオラクルに対して検証されている: torch per-stage golden → f32 TS ポート (≤2.3e-6) → graph parity で RoPE permutation を証明 (<1e-4) → GPU ids が CPU q8 oracle と bit 一致 → WAV が独立デコードと 9.9e-5。エンコーダは**リサンプラ差を分離**して正典を明示 (golden 比 6.0e-4 には Chrome と torchaudio の差 4.4e-2 が乗るので、同一入力の CPU 参照比 8.1e-7 が正典)。

int8 の代償も計測済み: greedy の軌道は f32 と step 6 で分岐する (近接値のフリップ)。以後も全 speech token で正常に eos に到達するので実害はないが、**GPU の検証を f32 列と突き合わせてはいけない**理由になっている。

### 残件

- **#120** sampled モードが ~25 tok/s (上流 sampler の全語彙 CPU softmax が律速)。greedy は 209 tok/s なので GPU 側の問題ではない
- **`src/` への統合が未起票** — spike は `src/` から何も export していない。`SynthesisEngine` の裏に入れるか、入れるとして 1.1GB のダウンロードと WebGPU 必須をどう扱うかは製品判断が要る。このエピックを閉じる前に決めたい
- 音の良し悪しは未評価。well-formed であることは自動で確認しているが、聞こえ方は人間の耳が要る (`cd spike/miotts && npm run serve`)

#105 (Sarashina, NonCommercial で停止) との比較では、こちらが license・サイズ・実測速度の全てで上回った。

**Comment by m96-chan** [comment]
## 顛末: 前提が外れて、答えが不要になった

**この ISSUE の完了条件は「MioCodec のデコーダを ONNX に出せるか」を軸に組まれていた。結論から言うと、その問いには答えていない。答える必要が無くなったから。**

### 何が起きたか

この ISSUE は正しく「**コーデックが死に場所**」を特定していた。MioCodec のデコーダは iSTFTHead で、ONNX エクスポータは複素テンソルを扱えない。Sarashina のエクスポートは `torch.stft` / `torch.istft` を実数の conv/matmul に置き換えて通していて、ここでも同じ作業が要る、と。**その診断は正確だった。**

外れたのは、その次の一手の前提のほう。「ONNX に出せなければ何も始まらない」は、**ONNX ランタイムを使うなら**正しい。実際には **WebGPU カーネルを直接書く道があった**。

上流に [web-xpu-ops](https://github.com/m96-chan/web-xpu-ops) を持っている以上、**無いものは作ればいい**。iSTFT が ONNX で表現できないなら、iSTFT のカーネルを書けばいい。実際そうした:

- `ops/stft` の逆変換 — ONNX が表現できなかったもの、そのもの
- `ops/matvec` の int8 経路 (`matvecQ8` / `packQ8`)、`ops/gqa`、`ops/rope`、`ops/conv_transpose`、`ops/group_norm` ほか計 16 op

**この ISSUE が「ブロッカー」と呼んでいたものは、上流に投げれば解けるものだった。** 迂回ではない — ONNX 経由よりむしろ素直で、途中の変換段が無いぶん、何が起きているかが見える。

### 完了条件に対して

- [ ] ~~MioCodec のデコーダが ONNX に出せるか、iSTFT の問題に具体的な答えが出ている~~
  **→ 問い自体が不要になった。** ONNX を経由しない。iSTFT の「具体的な答え」は「カーネルを書いた」

- [x] **日本語の発話が推定値ではなく実測値で 1 つ存在する** — 満たされた。RTX 5090、公開 API 経由、実重み:

  ```
  load:  1.2 s
  voice: 128 dims / 5.9 s (24.9 秒のクリップの前 20.00 秒)
  speak: 3.60 秒の音声を 2.08 秒 → RTF 0.58
  audio: peak 0.916 / RMS 0.134 / zero-crossing 2274 per s
  ```

  この ISSUE が引用していた上流の RTF 0.055-0.065 (vLLM, RTX 5090) は**比較できない数字**だと自分で書いている。こちらは**ブラウザで動く WebGPU の実測**なので、桁が違うのは当然で、意味が違う。

- [x] **実スコープの後続 ISSUE が存在する** — #116 (デコーダ) / #118 (TTS) / #121 (声クローン) / #129 (sampler) / #131 (Pages 公開) / #124 (`src/` 統合) が切られ、**全部マージ済み**。

### 到達点

**https://voxshot.m96-chan.dev/mio/ で稼働中。** 日本語テキスト → 音声と、参照音声からのゼロショット声クローンが、全部ブラウザ内で動く。

そして #124 で `voxshot/miotts` として**公開 API になった**:

```ts
import { VoxShot } from "voxshot";
import { createMioTtsEngine } from "voxshot/miotts";

const engine = await createMioTtsEngine({ weights: yourWeightSource });
const tts = await VoxShot.create({ engine, device: "webgpu" });
```

`SynthesisEngine` の裏に入っているので、この ISSUE が書いた「`embed()` と `synthesize()` でちょうど過不足ない」という見立ては**そのまま実現した**。24 kHz が `ChatterboxEngine` と一致するという指摘も、そのまま効いている。

派生物も公開した (ローカルの検証済みバイナリと sha256 一致):
- `m96-chan/MioTTS-0.6B-q8-webgpu` (apache-2.0, 611 MB)
- `m96-chan/MioCodec-encoder-webgpu` (mit, 117 MB)

### この ISSUE から学べること

**「ブロッカー」を正しく特定することと、それが本当にブロッカーかを判定することは別の作業。** この ISSUE は前者を正確にやった上で、後者を「ONNX 以外の道は無い」という暗黙の前提のまま通してしまっている。前提が書かれていなかったので、疑う対象にもならなかった。

自前の op ライブラリを持っているなら、**「エクスポートできない」は「まだ書いていない」と読み替えられる**。

### 残っているもの

- **#135** — QAT による 4bit 量子化で Web 配信サイズを半分にできないか (LM 583 MB → 約 292 MB)
- **#125** — 小さい seed で最初のトークンが実質 greedy になる
- **#115** — Irodori 調査
- chunk 間の韻律 (chunk ごと独立生成なので途切れる。LM が speech token に対して自己回帰である以上、前の chunk を文脈に与えれば繋がる可能性はあるが、参照サーバがそれをやっていないので訓練されているか不明 — #124 のコメントに保留として記録)

以上をもって完了とする。
